firmware/application/tone_key.cpp

```diff
@@ -114,10 +114,58 @@ std::string tone_key_string(tone_index index) {
     return tone_keys[index].first;
 }
 
-std::string tone_key_string_by_value(uint32_t value) {
-    return tone_key_string(tone_key_index_by_value(value));
+// Return string showing frequency only from specific table index
+std::string tone_key_value_string(tone_index index) {
+    if (index < 0 || (unsigned)index >= tone_keys.size())
+        return std::string("");
+    auto f = tone_keys[index].second;
+    return to_string_dec_uint((uint32_t)f) + "." + to_string_dec_uint((uint32_t)(f * 10) % 10);
+}
+
+// Return variable-length string showing CTCSS tone from tone frequency
+// Value is in 0.01 Hz units
+std::string tone_key_string_by_value(uint32_t value, size_t max_length) {
+    static uint8_t tone_display_toggle{0};
+    static uint32_t last_value;
+    tone_index idx;
+    std::string freq_str;
+
+    // If >10Hz difference between consecutive samples, it's probably noise, so ignore
+    if (abs(value - last_value) > 10 * 100) {
+        last_value = value;
```

Comment:
You're doing this unconditionally on 139, needed?

Reply:
Yes.  This is before a return statement and I need to update last_value whether it returns or not, but the if statement above uses last_value so I can't set it before the if statement unless perhaps I save off the result of the subtraction in another variable...